### PR TITLE
Allow FormSpec elements to be focused with `set_focus`

### LINKED ISSUE #8029: Allow formspecs to optionally specify the element to be focused
##### Issue type

 - Feature request

##### Summary

Currently, the first element of a formspec is highlighted, but sometimes the first element to be defined isn't the primary element. I suggest allowing formspecs to optionally specify the field to be focused.

e.g.
```lua
minetest.show_formspec(name, "test", [[
	size[6,7]
	field[4,1.25;2,1;secondary;Secondary;]
	button_exit[1,1;2,1;primary;Primary]
	set_focus[primary]
]])
```

### PR DESCRIPTION
**Pull request by v-rob**
This allows you to specify a FormSpec element to set the focus of with `set_focus[<name>;<always set>]`.  Closes #8029.

## Documentation from `lua_api.txt`

<blockquote>

### `set_focus[<name>;<always set>]`

* Sets the focus to the element with the same `name` parameter.
* **Note**: This element must be placed before the element it focuses.
* `always set` (optional, default `false`): If false, this sets the focus to the
  element specified only if the formspec is not being reshown. For instance, if
  the formspec with the name "test:example" is shown, the focus will be set, but
  if the same formspec is shown again consecutively, the focus will be set to
  the last thing the user focused before reshowing. If true, the focus will always
  be set in any circumstance.
* The following elements have the ability to be focused:
    * checkbox
    * button
    * button_exit
    * image_button
    * image_button_exit
    * item_image_button
    * table
    * textlist
    * dropdown
    * field
    * pwdfield
    * textarea
    * scrollbar
</blockquote>

Funny thing: Everything needed to make this PR was already implemented but unused.  I simply exposed this functionality to the API.

Note: Tabheaders and inventory lists aren't on this list because they aren't focusable in the first place -- try tabbing to either one.  It doesn't work.  That is work for another PR since I have no idea how to fix that.

## To do

This PR is Ready for Review.

## How to test

OK, to test.  What I did was to have every focusable element in a formspec all together.  Then, I created a copy of every one but with a new ID, and I tried focusing every one of these secondary elements.  Why second?  Because earlier elements are focused by default without `set_focus` according to some complicated rules, so we have to check that `set_focus` bypasses these.  Note that `button_exit` and `image_button_exit` work exactly the same as their non-exit counterparts, so they aren't included.

So, to test, use this formspec and change the `set_focus` to any secondary element, i.e. anything with a `-2` at the end of its ID and then press space or arrows, depending on the element.  If you remove the `set_focus`, `pwdfield-1` should be focused.  Also try testing the second parameter.

```lua
"formspec_version[3]"..
"size[9,11]"..
"set_focus[textlist-2]"..

"checkbox[1,0.5;checkbox-1;checkbox-1]"..
"button[1,1;3,0.7;button-1;button-1]"..
"image_button[1,2;3,0.7;air.png;image_button-1;image_button-1]"..
"item_image_button[1,3;3,0.7;air;item_image_button-1;item_image_button-1]"..
"table[1,4;3,0.7;table-1;table-1,-;1]"..
"textlist[1,5;3,0.7;textlist-1;textlist-1,-;1]"..
"dropdown[1,6;3,0.7;dropdown-1;dropdown-1,-;1]"..
"field[1,7;3,0.7;field-1;field-1;field-1]"..
"pwdfield[1,8;3,0.7;pwdfield-1;pwdfield-1]"..
"textarea[1,9;3,0.7;textarea-1;textarea-1;textarea-1]"..
"scrollbar[1,10;3,0.7;;scrollbar-1;scrollbar-1;0]"..

"checkbox[5,0.5;checkbox-2;checkbox-2]"..
"button[5,1;3,0.7;button-2;button-2]"..
"image_button[5,2;3,0.7;air.png;image_button-2;image_button-2]"..
"item_image_button[5,3;3,0.7;air;item_image_button-2;item_image_button-2]"..
"table[5,4;3,0.7;table-2;table-2,-;1]"..
"textlist[5,5;3,0.7;textlist-2;textlist-2,-;1]"..
"dropdown[5,6;3,0.7;dropdown-2;dropdown-2,-;1]"..
"field[5,7;3,0.7;field-2;field-2;field-2]"..
"pwdfield[5,8;3,0.7;pwdfield-2;pwdfield-2]"..
"textarea[5,9;3,0.7;textarea-2;textarea-2;textarea-2]"..
"scrollbar[5,10;3,0.7;;scrollbar-2;scrollbar-2;0]"
```

### CODE REVIEW

**Comment by SmallJoker** [comment]
How about `GUIFormSpecMenu::setFocus()`? That function is designed for exactly this purpose. As a side note: it will also preserve the focus on formspec updates.

**Comment by v-rob** [comment]
That function literally does `m_focused_element = elementname`, which is exactly the same as mine.  So, I can change it, but it won't change any functionality.  Mine still preserves focus across updates.

**Comment by Df458** [comment]
I think the main reason to use `GUIFormSpecMenu::setFocus()` instead of calling the code inside directly is for future-proofing. If focus isn't always set through the same mechanism, then later features/fixes/refactoring will have a significantly higher chance of producing bugs since this place will likely be overlooked.

**Comment by v-rob** [comment]
I take it back: I did have a reason to not use setFocus (I thought I must have, that's doing things late at night for you).  Formspec focus for the elements is in `parserData`, not `m_focused_element`.  Mine still preserves across updates because the whole form is reparsed.

So, if I want to change this (which I should for good future code), I will have to remove the variable from `parserData` and change all the references to it, and then keep a variable that tracks when the form is resized (same form) vs if a new form is being shown (which conflicts with a future PR I'm currently working on, but oh well).

**Comment by v-rob** [comment]
Wow, that must be a world record.  I have to rebase already.  But it's because of the new `GUIInventoryList`, so I'm wondering:  Can I set the focus of inventory lists now?  @DS-Minetest

**Comment by v-rob** [comment]
OK, everything's done now.  This makes sure that focus and table states are preserved _only_ when the same formspec is reshown (on window resize, mainly).  Why tables?  Because they share the same code as keeping focus, so making another PR feels redundant.

**Comment by v-rob** [comment]
Ugh.  No, everything's _not_ done.  I forgot the case where a mod reshows the same formspec with only minor changes, and how I've done it, the focus won't be preserved from where the user set it (assuming the mod doesn't explicitly set it with `set_focus`).  But the way it was before might accidentally set the focus where it shouldn't.  So, I have to check if the form name is the same instead.

**Comment by v-rob** [comment]
11 force pushes later, I've fixed everything (finally).  Check the first post for all the details because I had to make quite a few changes.  This is now ready for review.

**Comment by DS-Minetest** [comment]
> Wow, that must be a world record. I have to rebase already. But it's because of the new `GUIInventoryList`, so I'm wondering: Can I set the focus of inventory lists now? @DS-Minetest

Setting focus to inventorylists should theoretically be possible, but users can only interact with inventorylists by clicking on them, therefor setting focus on them wouldn't be useful.

**Comment by v-rob** [comment]
> Btw. it might make sense to allow using `set_focus` without a name by placing the `set_focus` element directly after (or before) the element to be focused, eg.:
> 
> ```
> size[...]
> button[...]
> set_focus[]
> field[...]
> ```

Interesting idea, but I don't think I'll add it.  I would have to add multiple booleans to do so, and then it wouldn't use the built in `setFocus`, so things outside of `GUIFormSpecMenu` wouldn't be able to set the focus to anything.  And, if you modify a formspec you are showing and decide that a different element should be focused, the entire element would have to be moved, which is trickier to code.

**Comment by runsy** [comment]
I missed this feature, really :-(

File: doc/lua_api.txt [hunk 1]
```diff
@@ -2396,8 +2396,8 @@ Elements
 * `name` fieldname data is transferred to Lua
 * `caption 1`...: name shown on top of tab
 * `current_tab`: index of selected tab 1...
-* `transparent` (optional): show transparent
-* `draw_border` (optional): draw border
+* `transparent`: show transparent
+* `draw_border`: draw border
```
Comment (SmallJoker) [inline doc/lua_api.txt:2400]: Unrelated change. These arguments are optional.

File: doc/lua_api.txt [hunk 2]
```diff
@@ -2409,8 +2409,8 @@ Elements
...
```
Comment (SmallJoker) [inline doc/lua_api.txt:2413]: Same here

File: doc/lua_api.txt [hunk 3]
```diff
@@ -2573,6 +2573,31 @@ Elements
...
+### `set_focus[<name>;<always set>]`
+
+* Sets the focus to the element with the same `name` parameter.
+* **Note**: This element must be placed before the element it focuses.
+* `always set` (optional, default `false`): If false, this sets the focus to the
+  element specified only if the formspec is not being reshown. For instance, if
+  the formspec with the name "test:example" is shown, the focus will be set, but
+  if the same formspec is shown again consecutively, the focus will be set to
+  the last thing the user focused before reshowing. If true, the focus will always
+  be set in any circumstance.
```
Comment (SmallJoker) [inline doc/lua_api.txt:2585]: This is kinda long. How about:
```
* `force` (optional): By default (`false`), focus is not applied for
  re-sent formspecs with the same name so that player-set focus is kept.
  `true` sets focus to a the specified element for each sent formspec.
```

File: src/gui/guiFormSpecMenu.cpp [hunk 4]
```diff
@@ -2687,6 +2687,26 @@ bool GUIFormSpecMenu::parseStyle(parserData *data, const std::string &element, b
...
 }
 
+void GUIFormSpecMenu::parseSetFocus(const std::string &element)
+{
+	std::vector<std::string> parts = split(element, ';');
+
+	if (parts.size() <= 2 ||
+		(parts.size() > 2 && m_formspec_version > FORMSPEC_API_VERSION))
+	{
+		if (!m_is_form_regenerated && // Never focus on resizing
```
Comment (SmallJoker) [inline src/gui/guiFormSpecMenu.cpp:2697]: How about early returns here? The code will be longer, but much easier to understand.

```C++
if (m_is_form_regenerated)
	return; // Never focus on resizing
```

File: doc/lua_api.txt (hunk 1 above)
Comment (v-rob) [reply doc/lua_api.txt:2400]: I don't know why I did this...

File: src/gui/guiFormSpecMenu.cpp [hunk 5]
```diff
@@ -655,7 +655,7 @@ void GUIFormSpecMenu::parseCheckbox(parserData* data, const std::string &element
 		auto style = getDefaultStyleForElement("checkbox", name);
 		e->setNotClipped(style.getBool(StyleSpec::NOCLIP, false));
 
-		if (spec.fname == data->focused_fieldname) {
+		if (spec.fname == m_focused_element) {
```
Comment (rubenwardy) [inline src/gui/guiFormSpecMenu.cpp:658]: Worth noting that I'm moving things _into_ RenderData in my formspec refactor, not out

Comment (v-rob) [reply src/gui/guiFormSpecMenu.cpp:658]: I added it here because the others wanted to use `setFocus`, which can't access `parserData`; trying to move stuff around probably wouldn't help the refactor much.
